modules/sdk/src/main/scala/org/tessellation/sdk/domain/snapshot/services/GlobalL0Service.scala

```diff
@@ -79,43 +108,124 @@ object GlobalL0Service {
-                    .min(singlePullLimit.map(nextOrdinal.value.value + _.value).getOrElse(lastOrdinal.value.value))
-
-                  nextOrdinal.value.value to lastOrdinalCap
-                }
-                .map(_.toList.map(o => SnapshotOrdinal(NonNegLong.unsafeFrom(o))))
-                .flatMap { ordinals =>
-                  (ordinals, List.empty[Hashed[GlobalIncrementalSnapshot]]).tailRecM {
-                    case (ordinal :: nextOrdinals, snapshots) =>
-                      l0GlobalSnapshotClient
-                        .get(ordinal)(l0Peer)
-                        .flatMap(_.toHashedWithSignatureCheck.flatMap(_.liftTo[F]))
-                        .map(s => (nextOrdinals, snapshots :+ s).asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
-                        .handleErrorWith { e =>
-                          logger
-                            .warn(e)(s"Failure pulling snapshot with ordinal=$ordinal")
-                            .map(_ => snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
-                        }
-
-                    case (Nil, snapshots) =>
-                      Applicative[F].pure(snapshots.asRight[(List[SnapshotOrdinal], List[Hashed[GlobalIncrementalSnapshot]])])
-                  }
-                }
-            }
-
-            pulled.map(_.asRight[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)])
+            for {
+              l0Peer <- globalL0ClusterStorage.getRandomPeer
+              lastOrdinal <- l0GlobalSnapshotClient.getLatestOrdinal.run(l0Peer)
+              pulled <- pullSnapshots(l0Peer, lastStoredOrdinal.next, lastOrdinal)
+            } yield pulled.asRight[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)]
           }
         }.handleErrorWith { e =>
           logger.warn(e)(s"Failure pulling global snapshots!") >>
             Applicative[F].pure(
               List.empty[Hashed[GlobalIncrementalSnapshot]].asRight[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo)]
             )
         }
+
+      def pullGlobalSnapshotsFromMajority
+        : F[Either[(Hashed[GlobalIncrementalSnapshot], GlobalSnapshotInfo), List[Hashed[GlobalIncrementalSnapshot]]]] =
+        maybeMajorityPeerIds.fold {
+          pullGlobalSnapshots
+        } { majorityPeerIds =>
+          lastGlobalSnapshotStorage.getOrdinal.flatMap {
+            _.fold {
+              pullLatestSnapshotFromMajority.map(_.asLeft[List[Hashed[GlobalIncrementalSnapshot]]])
+            } { lastStoredOrdinal =>
+              for {
+                snapshotData <- getLatestMajoritySnapshotData(majorityPeerIds)
+                l0Peers <- globalL0ClusterStorage.getPeers
+                lastOrdinal = snapshotData.ordinal
+                majorityPeer <- getPeerInTheMajority(lastOrdinal, snapshotData.hash, l0Peers)
+                pulled <- pullSnapshots(majorityPeer, lastStoredOrdinal.next, lastOrdinal)
```

Comment:
I think I would run a check on the downloaded data - we can't be sure that the data will match the declared hash. We should assume the node we pull from may be malicious, so zero trust policy. So what I would do is I would check that the hash of the last downloaded snapshot is the hash we determined as majority and then that all the pulled snapshots form a chain in regards to that majority snapshot - should then be sure that the data is right.

Reply:
Yeah, this should be done -- the hash verification is the only thing we need here for security.